### PR TITLE
Add tasks to support conversion of topic tags into topic links

### PR DESCRIPTION
**Pull request by mobaig**
As part of the move towards using content IDs to uniquely identify
content across gov.uk, SubscriberLists need to have their existing tags
converted from slugs to content IDs.

This commit adds rake tasks to:
- Query the Content Store with a slug and write the returned content ID
  to the corresponding SubscriberList links hash.
- Delete SubscriberLists that don't return a content ID from the Content
  Store.

Trello: https://trello.com/c/ZD65QUBN/326-email-alerts-based-on-links-hash

### CODE REVIEW

**Comment by mobaig** [comment]
Could do with review from @elliotcm and @rboulton in particular.

File: lib/tasks/links_migration/link_migrator.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        # Only one base path is ever specified in topic tags
+        subscriber_list.tags[:topics].first
+      end
+
+      def relevant_subscriber_lists
+        @relevant_subscriber_lists ||= SubscriberListQuery.new(query_field: :tags)
+          .subscriber_lists_with_key(:topics)
+      end
+
+      def namespaced(base_path)
```
Comment (elliotcm) [inline lib/tasks/links_migration/link_migrator.rb]: Technically this is returning the base path.  The argument to this method is the slug.

File: lib/tasks/links_migration/link_migrator.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+module Tasks
+  module LinksMigration
+    class LinkMigrator
+      def populate_topic_links
+        relevant_subscriber_lists.each do |sl|
+          base_path = base_path_from(sl)
+          content_item = Services.content_store.content_item(namespaced(base_path))
+          if content_item && content_item.content_id.present?
```
Comment (elliotcm) [inline lib/tasks/links_migration/link_migrator.rb]: Under what circumstances would we not have a content ID here, and is skipping them the right thing to do?

File: lib/tasks/links_migration/link_migrator.rb [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            puts "Updating links in SubscriberList #{sl.id}"
+            sl.update(links: {topics: [content_item.content_id]})
+          end
+        end
+      end
+
+      def destroy_non_matching_subscriber_lists
+        relevant_subscriber_lists.each do |sl|
```
Comment (elliotcm) [inline lib/tasks/links_migration/link_migrator.rb]: Using low-meaning variables is discouraged in [the styleguide](https://github.com/alphagov/styleguides/blob/master/ruby.md) when not on a single line.  Perhaps `list` here?

**Comment by elliotcm** [comment]
Generally looks good.  Having we checked that there's no subscribers to these phantom subscriber lists before we delete them?

File: lib/tasks/links_migration/link_migrator.rb (hunk 1 above)
Comment (mobaig) [inline lib/tasks/links_migration/link_migrator.rb]: :+1: will update...

File: lib/tasks/links_migration/link_migrator.rb (hunk 2 above)
Comment (mobaig) [inline lib/tasks/links_migration/link_migrator.rb]: Redirects.

Good shout though. Both methods were originally one thing, but in the current form of this class it probably makes sense to fail loudly if this is the case - the idea being the cleanup task below should be run first. Will update...

File: lib/tasks/links_migration/link_migrator.rb (hunk 3 above)
Comment (mobaig) [inline lib/tasks/links_migration/link_migrator.rb]: Updating...

**Comment by mobaig** [comment]
@elliotcm have reworked the approach here a bit. Key changes are detailed in the latest commit - would appreciate another look when you're able.

File: lib/tasks/links_migration/link_migrator.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+          end
+
+          if content_item.content_id.blank?
+            raise DodgyBasePathError, <<-ERROR.strip_heredoc
+              No content ID found for #{base_path_from(list)},
+              run `bundle exec rake links_migration:report_non_matching`
+              and fix these cases before continuing migration.
+            ERROR
+          end
+
```
Comment (jackscotti) [inline lib/tasks/links_migration/link_migrator.rb]: I thought content-store was checking for the validity of UUIDs, but it's not.
https://github.com/alphagov/content-store/blob/master/app/models/content_item.rb#L30

Do you think is worth checking for it here?

Comment (jackscotti) [inline lib/tasks/links_migration/link_migrator.rb]: I was wrong. It's doing it here: https://github.com/alphagov/content-store/blob/master/app/models/content_item.rb#L216

**Pull request merged**